Update colour coding of DisputeStatusChip based on dispute status and urgency

Jinksi opened the pull request:
Fixes #6722 

#### Changes proposed in this Pull Request

<!--
Title: A descriptive, yet concise, title.
-->

<!--
Description: Write a brief summary about this PR. As you compose your summary, consider each of these questions and address them if appropriate. Why is this change needed? What does this change do? Were there other solutions you considered? Why did you choose to pursue this solution? Describe any trade-offs you might have had to make.
-->

<!--
Questions for PR author:
- How can this code break?
- What are we doing to make sure this code doesn't break?
-->

<!--
Images or gifs: Include before and after screenshots or gifs/videos when it makes sense.
-->

This PR changes the colour of the Dispute Status Chip based on the following criteria:

* red = needs response within 72h
* yellow = needs response > 72h
* green = won
* blue = in progress / under review
* grey = lost


This Dispute Status Chip is rendered in the following locations, each with updated colours:

**Disputes List**
<img width="300" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/5a58f236-d00a-4477-a3ae-0b40166abe67"  />

**Dispute Details**
<img width="400" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/0d6b0cf4-be08-46fc-bb9e-85d9335695a2">

**Transaction Details**
<img width="400" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/120f3fa1-ec41-4333-9d49-f845856a9ee2">



#### Testing instructions

<!--
Testing instructions: How should this be tested and how can a reviewer test the end-user functionality? Are there known issues that you plan to address in a future PR? Are there any side effects that readers should be aware of?
-->

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Ensure your store has multiple disputes in various states, e.g. Lost, Won, Needs Response.
	* Possible statuses: `warning_needs_response`, `warning_under_review`, `warning_closed`, `needs_response`, `under_review`, `charge_refunded`, `won`, `lost`
	* Submit dispute evidence with `winning_evidence` in the `Additional Details` field to create a `Won` dispute. [See wiki](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows#winning-dispute).
* Navigate to Payments → Disputes. Ensure the colour of the chip under the **Status** column match the criteria listed above.
* Click on a dispute to view the Dispute Details screen. Ensure the status chip colour matches the criteria listed above.
* Click on the charge to view the Transaction Details screen. Ensure the status chip colour matches the criteria listed above.

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

<!--
Make sure you edit the page for the current release when adding testing instructions.
We often create a blank page ahead of time for the next release.
If this PR need not be QA tested, edit to 'QA Testing Not Applicable'
-->

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [x] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

brucealdridge committed:
Add action column with respond button to dispute list view

brucealdridge committed:
Right align action column

shendy-a8c committed:
Merge branch 'develop' into fix/6667-disputes-listing-add-respond-button

shendy-a8c committed:
Fix changelog so it does not add an entry.

shendy-a8c committed:
Make 'See details' button looks like proposed design.

Jinksi committed:
Merge branch 'develop' into fix/6667-disputes-listing-add-respond-button

Jinksi committed:
Use shared disputeAwaitingResponseStatuses variable

Jinksi committed:
Use client-side navigation on action button click

Jinksi committed:
Add WCPay tracks event for action button click

shendy-a8c committed:
Update snapshot.

Jinksi committed:
DisputeStatusChip : use red colour when urgent

Jinksi committed:
Migrate DisputeStatusChip to TypeScript

Jinksi committed:
Simplify isUrgent logic in DisputeDetails component

Jinksi committed:
Add 'success' type to Chip component

Jinksi committed:
Update dispute status chip mappings

Jinksi committed:
Merge branch 'develop' into fix/6722-dispute-list-status-chip-colours

botwoo commented:
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `6723` or branch name `fix/6722-dispute-list-status-chip-colours` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&branches.woocommerce-payments=fix/6722-dispute-list-status-chip-colours) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 3ba247c29d38f7704e3817ef06720bef4593e1ac
- Build time: 2023-07-13 02:03:04 UTC

_Note: the build is updated when a new commit is pushed to this PR._

Jinksi committed:
Migrate DisputeStatusChip tests to TypeScript

Jinksi committed:
Migrate Chip component to TypeScript

Jinksi commented on client/components/chip/test/index.test.tsx:
Refactored this test file to TypeScript. No changes to logic expected.

Jinksi commented on client/components/dispute-status-chip/test/index.test.tsx:
Refactored this test file to TypeScript. No changes to logic expected.

Jinksi commented on client/components/chip/style.scss:
Note: all of these background colours (e.g. `$studio-blue-5`) are darker than the Disputes Flows designs in Figma (which uses e.g. `$studio-blue-0` ).

I will confirm with design if we want to change these WCPay-wide and open a new issue if relevant.

<img width="602" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/45fd671c-74b4-496f-bf8d-3342534f9246">

Jinksi replied:
Asked in Slack: p1689057662945359-slack-C03LA109BKM

Jinksi replied:
Will address in a new PR.

Jinksi committed:
Merge branch 'develop' into fix/6722-dispute-list-status-chip-colours

Jinksi committed:
Update isDisputeUrgent docblock

Jinksi committed:
Create shared dispute isDueWithin util function

Jinksi committed:
Fix test case that was previously inaccurate

Jinksi commented on client/disputes/utils.ts:
```diff
@@ -0,0 +1,46 @@
+	dueBy: CachedDispute[ 'due_by' ] | EvidenceDetails[ 'due_by' ];
+	days: number;
+}
+export const isDueWithin = ( { dueBy, days }: IsDueWithinProps ): boolean => {
```
Ideally, use this shared util function throughout disputes codebase. Will handle this in followup PR.

Jinksi committed:
Revert Chip TS migration

Jinksi committed:
Revert migrate DisputeStatusChip to TS

Jinksi committed:
Use util function for dispute needing response statuses

Jinksi committed:
Improve docblock for isDueWithin function

Jinksi committed:
Update snapshots with default `warning` level for disputes needing response

Jinksi commented on client/components/dispute-status-chip/mappings.ts:
This file contains the core of the changes, aside from the urgent disputes needing a response = red.

Needs response = 🟡 Yellow
`needs_response`: warning
`warning_needs_response`: warning

In progress = 🔵 Blue
`warning_under_review`: primary
`under_review`: primary

Lost = ⚪️ Grey
`warning_closed`: light
`charge_refunded`: light
`lost`: light

Won = 🟢 Green
`won`: success

Jinksi commented on client/components/chip/style.scss:
```diff
@@ -12,6 +12,11 @@
 		color: $studio-blue-80;
 	}
 
+	&.chip-success {
```
Added new chip type/colour: success = green.

Jinksi committed:
Revert code quality changes to simplify PR

Jinksi committed:
Revert util function creation to simplify PR

Jinksi committed:
Revert changes to dispute-tasks to simplify PR

Jinksi committed:
Simplify and fix function docblock and types

Jinksi committed:
Merge branch 'develop' into fix/6722-dispute-list-status-chip-colours

Jinksi committed:
Add changelog entry

Jinksi committed:
Merge branch 'develop' into fix/6722-dispute-list-status-chip-colours

Jinksi committed:
Merge branch 'develop' into fix/6722-dispute-list-status-chip-colours

shendy-a8c committed:
Merge branch 'develop' into fix/6722-dispute-list-status-chip-colours

shendy-a8c commented on changelog/fix-6722-dispute-list-status-chip-colours:
```diff
@@ -0,0 +1,4 @@
+Significance: minor
+Type: update
+
+Improved communication of dispute statuses through updated status chip colors.
```
In https://github.com/Automattic/woocommerce-payments/issues/6355, there is a task "Add Changelog entry for this epics changes", so I think for this PR, no need to add changelog entry as it'd be done in a separate PR.

Jinksi replied:
Ah, that's right. It's so automatic to add a changelog 😆

Jinksi replied:
🤦 I didn't change this before merge. I've added this to the changelog task item on the epic #6355

shendy-a8c reviewed (commented):
There is a [local function named isDueWithin()](https://github.com/Automattic/woocommerce-payments/blob/fix/6722-dispute-list-status-chip-colours/client/overview/task-list/dispute-tasks.ts#L28). Do you think that needs to be removed and replaced with this new `isDueWithin()` from `client/disputes/utils.ts`?

shendy-a8c commented on client/disputes/utils.ts:
```diff
@@ -0,0 +1,47 @@
+import type { CachedDispute, EvidenceDetails } from 'wcpay/types/disputes';
+
+interface IsDueWithinProps {
+	dueBy?: CachedDispute[ 'due_by' ] | EvidenceDetails[ 'due_by' ];
```
I don't see any `isDueWithin()` call that does not pass `dueBy`.

```suggestion
	dueBy: CachedDispute[ 'due_by' ] | EvidenceDetails[ 'due_by' ];
```

Jinksi replied:
Thanks!

shendy-a8c commented on client/disputes/utils.ts:
```diff
@@ -0,0 +1,47 @@
+	}
+
+	const now = moment().utc();
+	const isWithinDays = dueByMoment.diff( now, 'days', true ) < days;
```
Got to ask again as per my understanding, 'within' is inclusive, so when diff is equal to the value of `days`, it should be true.

```suggestion
	const isWithinDays = dueByMoment.diff( now, 'days', true ) <= days;
```

Jinksi replied:
I agree – let's go with within is `<=`.

Updated this logic to your suggestion `<=` and made the unit tests explicit about this in 3ba247c.

Jinksi commented:
> There is a [local function named isDueWithin()](https://github.com/Automattic/woocommerce-payments/blob/fix/6722-dispute-list-status-chip-colours/client/overview/task-list/dispute-tasks.ts?rgh-link-date=2023-07-13T01%3A19%3A28Z#L28). Do you think that needs to be removed and replaced with this new isDueWithin() from client/disputes/utils.ts?

I initially has moved this function to `client/disputes/utils.ts` but I reverted this to keep this PR streamlined, especially due to the 6.2 deadline.

<img width="395" alt="image" src="https://github.com/Automattic/woocommerce-payments/assets/3147296/fba43f55-2dc3-49a4-8bd8-4afc434c6ecb">

I was planning to clean up duplicated logic calculating due_by times in a PR next week. What do you think?

shendy-a8c reviewed (approved):
Looks good and tested well! I left some feedbacks but none is crucial that should block the merge.

shendy-a8c commented:
> I was planning to clean up duplicated logic calculating due_by times in a PR next week. What do you think?

Sounds good.

Jinksi committed:
Remove inaccurate optional type

Co-authored-by: Shendy <73803630+shendy-a8c@users.noreply.github.com>

Jinksi committed:
Update isDueWithin logic to be inclusive of exact difference

htdat commented:
👋 @Jinksi and @shendy-a8c - it looks like this PR is causing this E2E test suite failure: 


> FAIL tests/e2e/specs/wcpay/merchant/merchant-disputes-submit-winning.spec.js (123.065 s)
> Disputes > Submit winning dispute
> ✕ should process and confirm a winning dispute (100010 ms)
> 
> ● Disputes > Submit winning dispute › should process and confirm a winning dispute

It appears in the last commit https://github.com/Automattic/woocommerce-payments/commit/3ba247c29d38f7704e3817ef06720bef4593e1ac here as well as since this PR was merged to `develop` in https://github.com/Automattic/woocommerce-payments/commits/develop

![Markup on 2023-07-14 at 10:24:01](https://github.com/Automattic/woocommerce-payments/assets/10045087/29f800b6-5d6e-4b71-bc8d-dbb0c072d3a7)

Jinksi commented:
Thanks @htdat - apologies for missing this. I hadn’t looked closely enough at the e2e tests since there had been unrelated tests failing all week 🤦‍♂️

I’ll get a fix ready.

Jinksi commented:
PR #6787 ready for review
